Fix find_current_module so that it can work in application bundles

astrofrog opened the pull request:
As part of work on making it possible to use astropy in application bundles (#8795) I had to re-use a fix @embray originally wrote in #960 but since that PR never got merged, I thought I'd open a dedicated PR for this fix with a test. Essentially the goal here is to make sure that ``find_current_module`` works even if ``inspect.getmodule`` doesn't work, which can happen if astropy is packaged up inside a bundle created by e.g. ``pyinstaller`` or ``py2app``.

pllim commented:
Do we really need to backport this that far to 2.x series? People still trying to bundle for Python 2? If not, it is easier to backport only to 3.x.

astrofrog commented:
No strong feelings about milestone, but it's possible people will still use LTS with e.g. pyinstaller.

astrofrog commented:
Ok, so there is a failure with one of the modeling doctests:

```
______________ [doctest] astropy.modeling.core._ModelMeta.rename _______________
201 
202         Creates a copy of this model class with a new name.
203 
204         The new class is technically a subclass of the original class, so that
205         instance and type checks will still work.  For example::
206 
207             >>> from astropy.modeling.models import Rotation2D
208             >>> SkyRotation = Rotation2D.rename('SkyRotation')
209             >>> SkyRotation
Differences (unified diff with -expected +actual):
    @@ -1,3 +1,3 @@
    -<class '__main__.SkyRotation'>
    +<class 'astropy.modeling.core.astropy.modeling.core.SkyRotation'>
     Name: SkyRotation (Rotation2D)
     Inputs: ('x', 'y')
/tmp/astropy-test-f_0hhe45/lib/python3.7/site-packages/astropy/modeling/core.py:209: DocTestFailure
```

What do we think is the correct behavior here?

nden commented:
I like the new output because it's clear the class comes from the modeling package. 
It's always going to be in the `modeling.core` module so I think this is also good.
I can't think of any disadvantages of having this path to the new class.

bsipocz commented:
Alternatively we can always use an ellipsis for that line.

pllim commented:
Change log also needs updating.

astrofrog commented:
Looks like this still needs some work on Windows. I'll investigate next week.

saimn commented:
`__main__.SkyRotation` makes more sense to me, this is a new class so it should inherit the module (or here `__main__`) where it is defined.
Also the duplicated module name in `astropy.modeling.core.astropy.modeling.core.SkyRotation` is a bit weird.

astrofrog commented:
@saimn - good catch about the duplicate path in the module name, not sure why I didn't see that. This was a real bug unrelated to the original intent of this PR, and is now fixed.

I think that for the doctest, the path to the class is correct since the doctest is in fact inside ``astropy.modeling.core`` - but I've now added a more extensive test that ensures that if a user runs a script, the path to the class will include ``__main__``, and indeed it does work correctly.

astrofrog commented on astropy/modeling/core.py:
```diff
@@ -227,13 +227,7 @@ def rename(cls, name):
-                new_cls.__qualname__ = name
-            else:
-                new_cls.__qualname__ = '{0}.{1}'.format(modname, name)
+        new_cls.__qualname__ = name
```
This was a bug - ``__qualname__`` should not include the module name (see [PEP 3155](https://www.python.org/dev/peps/pep-3155/))

astrofrog commented:
@nden - can you review this updated version?

bsipocz commented:
(I've asked the re-review from @pllim, too in the hope of not having the green ticks on the right hand side would also mean that the "Approved" status will also disappear when looking at the PR list. Apparently that's not the case 😞  )

saimn commented:
> I think that for the doctest, the path to the class is correct since the doctest is in fact inside astropy.modeling.core

Yes indeed, though it depends on how you execute the doctests (`python -m doctest` or `import doctest; doctest.testmod()` will give different outputs). With pytest and the current directory structure, pytest imports the module from the package so I think that the current output (with your PR) is good.

pllim commented:
Might need a rebase? The test failure doesn't make much sense:
```
ERROR: Python >=3.6 is required by astropy
```

pllim commented:
> Looks like this still needs some work on Windows.

@astrofrog , is the Windows work all done?

bsipocz commented:
wow, that is weird. Why does it already pick up the 3.6+ requirement, but not the changes to the travis matrix?

nden commented:
@astrofrog @saimn What do you think is sensible in the case when a model is defined in another package, for example:

```
from gwcs.selector import LabelMapper
lm = LabelMapper.rename('LLL')
```
With the current PR and current master
```
lm                                                                                                                                            
Out[3]: 
<class '__main__.LLL'>
Name: LLL (LabelMapper)
Inputs: <property object at 0x7f42e517d048>
Outputs: ('label',)
```

I don't have a use case for this method but (perhaps naively) I would expect to see a reference to the `gwcs.selector` module.

If I change the `rename` method to get the module name by
``` 
new_cls = type(name, (cls,), {})
mod = inspect.getmodule(cls)
```

I get instead
``` lm                                                                                                                                            
Out[3]: 
<class 'gwcs.selector.LLL'>
Name: LLL (LabelMapper)
Inputs: <property object at 0x7f78cb028ef8>
Outputs: ('label',)
```

And for the doctest example
```SkyRotation                                                                                                                                   
Out[6]: 
<class 'astropy.modeling.rotations.SkyRotation'>
Name: SkyRotation (Rotation2D)
Inputs: ('x', 'y')
Outputs: ('x', 'y')
Fittable parameters: ('angle',)
```
That said I don't know if this would prevent it from running in a bundle (the original issue being solved). 
I'm mostly trying to understand the issue.

astrofrog commented:
@nden - yes I think that for models defined in other packages we should indeed do better than just ``__main__``. Since that occurs in both master and this PR, I suggest that might be best dealt with in a separate issue/PR though?

astrofrog commented:
Oops I force pushed an older version of the branch. I'll have to see if I can recover the original version. I'm traveling so this might have to wait until after, as I did this on my desktop.

bsipocz commented:
@astrofrog - I guess using the hash above you might be able to recover it from your fork?

astrofrog commented:
Ok so this wasn't super easy as there's no way to actually just fetch the commits directly - instead I had to download each original commit as a patch and apply it with ``git am``, e.g.

```
wget https://github.com/astropy/astropy/commit/08d029e6b85be197590d43811e59acae7ea06d70.patch
git am 08d029e6b85be197590d43811e59acae7ea06d70.patch
```

This should now be back to the latest version.

pllim commented:
@astrofrog , just to confirm, do you still have to "check something on Windows" or not?

astrofrog commented:
@nden - no, I did it and fixed the issue, AppVeyor is passing :)

astrofrog commented:
I'll go ahead and merge since this was approved!

bsipocz commented:
I get all kind of errors with this on LTS, so given that no strong preference was shown above, would incline not to backport. One example:
```
_______________________________________ test_zero_degree_polynomial[Chebyshev2D] _______________________________________

cls = <class 'astropy.modeling.polynomial.Chebyshev2D'>
Name: Chebyshev2D (OrthoPolynomialBase)
Inputs: ('x', 'y')
Outputs: ('z',)
Fittable parameters: <property object at 0x12abfa778>

    @pytest.mark.parametrize('cls', (Polynomial1D, Chebyshev1D, Legendre1D,
                                     Polynomial2D, Chebyshev2D, Legendre2D))
    def test_zero_degree_polynomial(cls):
        """
        A few tests that degree=0 polynomials are correctly evaluated and
        fitted.
    
        Regression test for https://github.com/astropy/astropy/pull/3589
        """
    
        if cls.n_inputs == 1:  # Test 1D polynomials
            p1 = cls(degree=0, c0=1)
            assert p1(0) == 1
            assert np.all(p1(np.zeros(5)) == np.ones(5))
    
            x = np.linspace(0, 1, 100)
            # Add a little noise along a straight line
            y = 1 + np.random.uniform(0, 0.1, len(x))
    
            p1_init = cls(degree=0)
            fitter = fitting.LinearLSQFitter()
            p1_fit = fitter(p1_init, x, y)
    
            # The fit won't be exact of course, but it should get close to within
            # 1%
            assert_allclose(p1_fit.c0, 1, atol=0.10)
        elif cls.n_inputs == 2:  # Test 2D polynomials
            if issubclass(cls, OrthoPolynomialBase):
                p2 = cls(x_degree=0, y_degree=0, c0_0=1)
            else:
                p2 = cls(degree=0, c0_0=1)
            assert p2(0, 0) == 1
            assert np.all(p2(np.zeros(5), np.zeros(5)) == np.ones(5))
    
            y, x = np.mgrid[0:1:100j, 0:1:100j]
            z = (1 + np.random.uniform(0, 0.1, x.size)).reshape(100, 100)
    
            if issubclass(cls, OrthoPolynomialBase):
                p2_init = cls(x_degree=0, y_degree=0)
            else:
                p2_init = cls(degree=0)
            fitter = fitting.LinearLSQFitter()
            p2_fit = fitter(p2_init, x, y, z)
    
>           assert_allclose(p2_fit.c0_0, 1, atol=0.10)
E           AssertionError: 
E           Not equal to tolerance rtol=1e-07, atol=0.1
E           
E           Mismatch: 100%
E           Max absolute difference: 1.
E           Max relative difference: 1.
E            x: array(0.)
E            y: array(1)

astropy/modeling/tests/test_polynomial.py:362: AssertionError
------------------------------------------------- Captured stderr call -------------------------------------------------
Python(78280,0x7fff971c3380) malloc: *** mach_vm_map(size=18446744071772295168) failed (error code=3)
*** error: can't allocate region
*** set a breakpoint in malloc_error_break to debug
init_dgelsd failed init
__________________________
```

astrofrog commented:
Oh fun. Agree on not backporting. Will you manually change the changelog section?

bsipocz commented:
Oh, never mind me, I realized that I used np 1.18.x for testing, so now trying the backport again. 

And yes, I'll manually clean up changelog inconsistencies before a release is done.

bsipocz commented:
(I still suspect that there are a bit too many absolute imports that may not work the way expected)

bsipocz commented:
Now run into only one failure of comparing unicodes to strings:
```
___________________________________________________ test_rename_path ___________________________________________________

tmpdir = local('/private/var/folders/dc/hsm7tqpx2d57n7vb3k1l81xw0000gq/T/pytest-of-bsipocz/pytest-268/test_rename_path0')

    def test_rename_path(tmpdir):
    
        # Regression test for a bug that caused the path to the class to be
        # incorrect in a renamed model's __repr__.
    
        assert repr(RENAMED_MODEL).splitlines()[0] == "<class 'astropy.modeling.tests.test_core.CustomGaussian'>"
    
        # Make sure that when called from a user script, the class name includes
        # __main__.
    
        env = os.environ.copy()
        paths = [os.path.dirname(astropy.__path__[0])] + sys.path
        env['PYTHONPATH'] = os.pathsep.join(paths)
    
        script = tmpdir.join('rename.py').strpath
        with open(script, 'w') as f:
            f.write(MODEL_RENAME_CODE)
    
        output = subprocess.check_output([sys.executable, script], env=env)
>       assert output.splitlines() == MODEL_RENAME_EXPECTED.splitlines()
E       assert ["<class 'ast...ssian'>", ...] == ["<class 'astr...ssian'>", ...]
E         At index 2 diff: "Inputs: (u'x',)" != "Inputs: ('x',)"
E         Use -v to get the full diff

astropy/modeling/tests/test_core.py:426: AssertionError
```

pllim commented:
Looks like `output.splitlines()` is spewing out unicode in PY2, so maybe something like this? 

```python
list(map(str, output.splitlines()))
```

Disclaimer: Just a theoretical guess from a glance; untested.
